core/vcpu.c

```diff
@@ -2772,13 +2800,26 @@ static void handle_cpuid_virtual(struct vcpu_t *vcpu, uint32_t a, uint32_t c)
             return;
         }
         case 0x80000008: {              // Virtual/Physical Address Size
+            struct vm_t *vm = vcpu->vm;
+            hax_list_head *list;
+            int vcpu_count = 0;
+
+            hax_mutex_lock(vm->vm_lock);
+            hax_list_for_each(list, (hax_list_head *)(&vm->vcpu_list)) {
+                vcpu_count++;
+            }
+            hax_mutex_unlock(vm->vm_lock);
+
             // Bit mask to identify the reserved bits in paging structure high
             // order address field
             physical_address_size = (uint8_t)state->_eax & 0xff;
             pw_reserved_bits_high_mask =
                     ~((1 << (physical_address_size - 32)) - 1);
 
             state->_ebx = state->_ecx = state->_edx = 0;
+
+            if (vcpu_count > 1)
+                state->_ecx |= vcpu_count - 1;
```

Comment:
According to Intel SDM, when the initial EAX value is 0x80000008, ECX is reserved to 0. Why set state->_ecx here?

Reply:
You are right, SDM says it's reserved to 0. I've done this since qemu does the same:
```
    case 0x80000008:
        /* virtual & phys address size in low 2 bytes. */
        if (env->features[FEAT_8000_0001_EDX] & CPUID_EXT2_LM) {
            /* 64 bit processor */
            *eax = cpu->phys_bits; /* configurable physical bits */
            if  (env->features[FEAT_7_0_ECX] & CPUID_7_0_ECX_LA57) {
                *eax |= 0x00003900; /* 57 bits virtual */
            } else {
                *eax |= 0x00003000; /* 48 bits virtual */
            }
        } else {
            *eax = cpu->phys_bits;
        }
        *ebx = env->features[FEAT_8000_0008_EBX];
        *ecx = 0;
        *edx = 0;
        if (cs->nr_cores * cs->nr_threads > 1) {
            *ecx |= (cs->nr_cores * cs->nr_threads) - 1;
        }
        break;
```